src/Firebase/AppCheck/AppCheckTokenVerifier.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Kreait\Firebase\AppCheck;
+
+use Firebase\JWT\CachedKeySet;
+use Firebase\JWT\JWT;
+use Kreait\Firebase\Exception\AppCheck\FailedToVerifyAppCheckToken;
+use Kreait\Firebase\Exception\AppCheck\InvalidAppCheckToken;
+use LogicException;
+use UnexpectedValueException;
+
+use function in_array;
+use function is_string;
+use function str_starts_with;
+
+/**
+ * @internal
+ */
+class AppCheckTokenVerifier
+{
+    private const APP_CHECK_ISSUER_PREFIX = 'https://firebaseappcheck.googleapis.com/';
+    private string $projectId;
+    private CachedKeySet $keySet;
+
+    public function __construct(string $projectId, CachedKeySet $keySet)
+    {
+        $this->projectId = $projectId;
+        $this->keySet = $keySet;
+    }
+
+    /**
+     * Verfies the format and signature of a Firebase App Check token.
+     *
+     * @param string $token the Firebase Auth JWT token to verify
+     *
+     * @throws FailedToVerifyAppCheckToken if the token could not be verified
+     * @throws InvalidAppCheckToken if the token is invalid
+     */
+    public function verifyToken(string $token): DecodedAppCheckToken
+    {
+        $decodedToken = $this->decodeJwt($token);
+
+        $this->verifyContent($decodedToken);
+
+        return $decodedToken;
+    }
+
+    /**
+     * @param string $token the Firebase App Check JWT token to decode
+     *
+     * @throws FailedToVerifyAppCheckToken if the token could not be verified
+     * @throws InvalidAppCheckToken if the token is invalid
+     */
+    private function decodeJwt(string $token): DecodedAppCheckToken
+    {
+        try {
+            $payload = (array) JWT::decode($token, $this->keySet);
```

Comment:
I'm not 100% sure, but I think we could do a

```php
/** @var DecodedAppCheckTokenShape $payload */
$payload = (array) JWT::decode($token, $this->keySet);
```

here, together with my comment on the `DecodedAppCheckToken` this should save us some code 🤞

Reply:
I'm not sure I follow this one. Are you saying the type-hint is all we need here and we shouldn't concern ourselves with the exception handling? I'm not sure I understand how static analysis is going to help us if the developer provides a string that isn't a valid token at runtime.